Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/SolverHandler.cs

```diff
@@ -310,6 +310,14 @@ private void Update()
 
         private void LateUpdate()
         {
+            if (updateSolversList)
+            {
+                IEnumerable<Solver> inspectorOrderedSolvers = GetComponents<Solver>().Intersect(solvers);
```

Comment:
How does Intersect() guarantee order? 

Also I don't think GetComponents work. this assumes all solvers will want to associate with this particular SolverHandler which is a bit dangerous. Like InBetween is an interesting scenario here. 

It also makes it difficult for inserting new solvers dynamically. If I can add a new solver dynamically, then it appears this will be put at the end of the list? But how would I insert that solver higher via GetComponents<>? Furthermore, if I was to register a solver with a solverhandler that is not on the sameobject...you could never insert it before other solvers.

Reply:
`solvers` will always be a list of only the solver that have registered, or have been set programmatically by something that wants to override the managed solvers.

`GetComponents` returns the list of all solvers in Insepctor order attached to this Gameobject. Interset() takes two lists and returns a list of elements that are in both lists, but keeping the order of the first list. See [Intersect<TSource>(IEnumerable<TSource>, IEnumerable<TSource>)](https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.intersect?view=netframework-4.8#System_Linq_Enumerable_Intersect__1_System_Collections_Generic_IEnumerable___0__System_Collections_Generic_IEnumerable___0__), specifically the example and remarks section:

> When the object returned by this method is enumerated, `Intersect` yields distinct elements occurring in both sequences in the order in which they appear in first.

So `inspectorOrderedSolvers` is the list of solvers attached to this gameobject in Inspector order, but only those in the `solvers` list.  Finally using `Union` we add the remaining solvers, that are not attached to this Gameobject, but are in the `solvers` list.

So the order will always be: Insepctor order of Solvers attached to this Gameobject that have been registered, then remaining Solvers not attached to this Gameobject, in the order of their appearance in `solvers` (By default order of `Register` calls).

If you dynamically add a Solver, `AddComponent` will add them at the end of the list of components. Though you can mix up the order and manually place it at another position. Currently this change will only take effect, when you do it right away when adding the solver, before LateUpdate runs during that frame. So maybe there should be a public API to trigger the Solver List Update. Also the `Solvers` setter maybe should trigger Solver List Update, so the Insepctor Order is guaranteed.

I think the underlying problem is, what shall the ordering of solver updates be? Inspector Order, Order of register call or programatically set order?  And further how to resolve conflicts due to the different ways to add solvers and manipulate the solver list. Like, default is Inspector Order, but what happens when I override the solver list? Does the list order take precedence, is that reflected back to the Inspector, as in the component order is changed? Or does the inspector takes precedence and the list order is changed internally to reflect the current Inspector, like it is now? 

Regarding Solvers that are not attached to the SolverHandler Gameobject, that is an edge case. That edge case didn't work with the previous implementation. Now it does, but the Inspector Ordering is hard to apply, like which Inspector takes precedence. I'm open for discussion as to how the ordering shall apply there. To create such edge case, registering the solver has to be done manually in code by the developer, though.

Reply:
@julianloehr-kg thanks for the great & detailed response. 

I did look at the Intersect docs but did not note the remarks. Good call out. Thank you.

I am going to file a bug that defining precedence/order for SolverHandler is a bit nuanced and problematic and will include all the items we discussed above. 

I'd rather get all your good changes here and fix the actual problem this PR was created for initially. We can do a separate PR to fix how priority can be easily defined for Solvers by consumers.